### DISCUSSION TITLE
Is there any way to get the predicted value? (predicted value & target value to draw the diagonal chart)

### DISCUSSION THREAD

**Discussion by ccccyf** [mir-group/nequip#496] [Q&A] [answered]
Wandb project just record the loss curve, but how to obtain the predicted values? I want to draw a diagonal graph.

**Comment by cw-tan** [comment, marked as answer]
You can use a [`test`](https://nequip.readthedocs.io/en/latest/guide/workflow.html#testing) with the [`TestTimeXYZFileWriter`](https://nequip.readthedocs.io/en/latest/api/callbacks.html#nequip.train.callbacks.TestTimeXYZFileWriter) to generate a `.xyz` file with both reference and predicted value on the test set (you must configure a test set to do this). Then, you can use this [script](https://github.com/mir-group/nequip/blob/main/examples/parity_plot.py) to make a parity plot. The description is outdated, but it works with this setup, just call it as `python parity_plot.py path/to/xyz_from_TestTimeXYZFileWriter`.

**Reply by ccccyf** [reply]
Thank you for your help! 
I need to write
```
    - _target_: nequip.train.callbacks.TestTimeXYZFileWriter
      out_file: ${hydra:runtime.output_dir}/test
```
part in the callbacks part of trainer.

**Reply by matr1x-1** [reply]
Is there a way to evaluate the validation or training set as well? For this we would need to get access to the exact split that results from our random seed choice.

**Reply by cw-tan** [reply]
@matr1x-1 , yes you would need to get access to the exact splits. We provide the automatic splits for convenience, but if metrics for those exact splits matter to you, we believe that users should perform those split themselves as a pre-processing step. The intended use case for the automatic splits is for stuff like models trained specifically for MD applications. If one just has a bunch of data collected, one could just do automatic train-val splits and "test" is done more on some actual benchmark tasks than just looking at test error metrics. The benchmark tasks could be things like energy-volume curves, simple MD for diffusion constants and RDFs, etc, which are often a lot more informative than test error metrics. In these cases, the data is likely just a pile of DFT frames sampled through whatever means, rather than a specially curated dataset for ML potential benchmarking purposes with predefined train-val-test splits.

Anyway, to answer your question specifically, you just have to pre-split your dataset, and re-register the train and val datasets as additional test datasets (the framework can handle multiple test datasets), and use the same XYZ writer callback.

**Reply by matr1x-1** [reply]
@cw-tan Makes sense. I was thinking about writing a custom callback but I guess the approach you mentioned is easier. Thanks!
